minetest/minetest
# Add physics modifier API

This prevents mods from conflicting by adding relative physics modifiers.

Modifiers can either be additive or multiplicative. Additions are done before multiplications.

Mods can still use the set physics override API, which will disable all modifiers.

Based on @raymoo's PR (#7269), but with the following changes:

* Rebased and improved code structure
* Allow unsetting set_physics_override
* Support for additive modifiers
* `get_effective_physics()`
* Move flags out of modifiers, into `set_physics_flags`
* devtest test mod

## Examples

```lua
-- Add multiplier for gravity
player:set_physics_modifier("planets:moon", {
	gravity = 0.16,
})

-- Add add modifier for speed
player:set_physics_modifier("potions:speed", {
	type = "add",

	speed = 0.5,
})

-- Remove modifier
player:set_physics_modifier("planets:moon", nil)

-- Different method is needed to set flags
player:set_physics_flags({
    sneak_glitch = false
})
```

## Documentation

* `get_effective_physics()`:
    * Returns a table of the physics factors after applying all overrides.
    * Table fields:
        * `speed`: multiplier to default walking speed value (default: `1`)
        * `jump`: multiplier to default jump value (default: `1`)
        * `gravity`: multiplier to default gravity value (default: `1`)
* `set_physics_modifier(id, modifiers)`:
   * Adds a physics modifier to the player under the given id.
   * Modifiers are resolved by first adding all `add` modifiers, and then multiplying all `multiply` modifiers.
   * `id`: A string to identify this modifier in the format "modname:modifier_name".
   * `modifiers`: either `nil` to delete the modifier, or a table with the same fields as `get_effective_physics` plus:
       * `type`: `multiply` or `add`. Default: `multiply`.
   * Examples:

     ```lua
     player:set_physics_modifier("mymod:moon", {
          gravity = 0.16,
     })

     player:set_physics_modifier("phystest:potion", {
         type = "add",
         speed = 0.5,
     })

     player:set_physics_modifier("phystest:potion", nil)
     ```

* `get_physics_modifiers()`: returns table of modifier id to modifier table, see above.
* `set_physics_flags(flags)`:
    * `sneak`: whether player can sneak (default: `true`)
    * `sneak_glitch`: whether player can use the new move code replications
      of the old sneak side-effects: sneak ladders and 2 node sneak jump
      (default: `false`)
    * `new_move`: use new move/sneak code. When `false` the exact old code
      is used for the specific old sneak behaviour (default: `true`)
* `get_physics_flags()`: returns the table given to `set_physics_flags`.
* `set_physics_override(override_table)`:
    * If `override_table` is a table:
        * Overrides the player's movement physics parameters and disables physics modifiers.
        * Can contain the same fields as `get_effective_physics` and `set_physics_flags`.
        * Setting physics flags here is DEPRECATED.
    * if `override_table` is nil, the use of physics modifiers will be re-enabled.
* `get_physics_override()`: returns the table given to `set_physics_override`.

## To do

- [x] Fix bug on overrides reset.
- [x] Add support for additive modifiers - these will be applied either before or after multipliers (TBD).
- [x] Consider flag behaviour.
- [x] Add`set_physics_flags()`.
- [x] Add deprecation log with flags in overrides

## How to test

See the /physics command in DevTest:

* `/physics` - list effective physics 
* `/physics override` - set physics overrides (will disable mods)
* `/physics reset` - reset physics overrides (will reenable mods)
* Modifiers:
    * `/physics moon` - add moon physics modifier
    * `/physics nomoon`
    * `/physics speed` - add speed physics modifier
    * `/physics nospeed`
    * `/physics potion` - adds speed physics modifier which is additive
    * `/physics nopotion`
    * `/physics jupiter` - add jupiter physics modifier
    * `/physics nojupiter`
    * `/physics invalid` - tests invalid type
* Flags:
	* `/physics glitch` - enable sneak glitch
	* `/physics noglitch` - disable sneak glitch
